oauthlib/oauth2/rfc6749/tokens.py

```diff
@@ -303,3 +301,47 @@ def estimate_type(self, request):
             return 5
         else:
             return 0
+
+
+class JWTToken(TokenBase):
+    __slots__ = (
+        'request_validator', 'token_generator',
+        'refresh_token_generator', 'expires_in'
+    )
+
+    def __init__(self, request_validator=None, token_generator=None,
+                 expires_in=None, refresh_token_generator=None):
+        self.request_validator = request_validator
+        self.token_generator = token_generator or random_token_generator
+        self.refresh_token_generator = (
+            refresh_token_generator or self.token_generator
+        )
+        self.expires_in = expires_in or 3600
+
+    def create_token(self, request, refresh_token=False, save_token=False):
+        """Create a JWT Token, using requestvalidator method."""
+
+        if callable(self.expires_in):
+            expires_in = self.expires_in(request)
+        else:
+            expires_in = self.expires_in
+
+        request.expires_in = expires_in
+
+        return self.request_validator.get_id_token(None, None, request)
+
+    def validate_request(self, request):
+        token = None
+        if 'Authorization' in request.headers:
+            token = request.headers.get('Authorization')[7:]
+        else:
+            token = request.access_token
+        return self.request_validator.validate_id_token(
+            token, request.scopes, request)
+
+    def estimate_type(self, request):
+        token = request.headers.get('Authorization', '')[7:]
+        if token.count('.') in (2, 4):
```

Comment:
Curious: When would there be 4 segment separators?

Also, there's no rule saying a normal access token can't contain dots of course. Wondering if you should include a check for `ey` or `.ey` here perhaps.

Reply:
I really don't remember... Will try to generate every accepted ID tokens by the specification to see a way to improve this tests

Reply:
@skion Found it!

* https://tools.ietf.org/html/rfc7516#section-9

and for access token no period is used:

 https://github.com/idan/oauthlib/blob/7a0783d356504867f4a4868717f6e5e7d1b90a10/oauthlib/common.py#L194-L203
https://github.com/idan/oauthlib/blob/7a0783d356504867f4a4868717f6e5e7d1b90a10/oauthlib/common.py#L29-L31

Reply:
Ah, JWE's! :+1: 

About the non-JWT character set: Yes, that's how oauthlib generates tokens, but another server might still include periods in them. And someone might use oauthlib as a *client* against a 3rd party server...

That said, I'm not sure what the best way is either, just that checking for `.`'s feels a bit shallow.

Reply:
When used as client oauthlib validate tokens before send it to 3rd party server?

It's guaranteed `ey` is added to every JWT token?

Can we let this check and see if it is a problem?

Reply:
> When used as client oauthlib validate tokens before send it to 3rd party server?

Exactly

> It's guaranteed ey is added to every JWT token?

I tried this yesterday and could not thing of a reason why a JWT would **not** start with `{"`, but haven't thoroughly checked it either.

> Can we let this check and see if it is a problem?

Just beware that you'll go wrong with the first [example token](https://tools.ietf.org/html/rfc6750#section-2.1) in the RFC :fearful:

Reply:
@skion I'm can't see other approach to solve this than validate the token at it early stage. What you think?

Some thing like:
```
>>> from jwcrypto import jwt
>>>jwt.JWT(jwt='token')
...
ValueError: Token format unrecognized

>>> jwt.JWT(jwt='t.oke.n')
...
InvalidJWSObject: Invalid JWS Object [Invalid format] {ValueError('Invalid base64 string',)}

>>> jwt.JWT(jwt='t.o.k.e.n')
...
InvalidJWEData: Invalid format {ValueError('Invalid base64 string',)}

>>> jwt.JWT(jwt='eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiYWRtaW4iOnRydWV9.TJVA95OrM7E2cBab30RMHrHDcEfxjoYZgeFONFh7HgQ')
<jwcrypto.jwt.JWT at 0x7f059df786d8>
```

Reply:
Yeah, that crossed my mind as well, but the tricky bit with it would be that you'd need to exclude all exceptions that might be raised for valid but expired or otherwise bad tokens.

And I also liked your PR in the way that the JWT logic itself was contained to the request validator, leaving the developer free to use whatever JWT library they might prefer.

Perhaps the following is a reasonable trade off that doesn't pull in a JWT lib as a dependency?

```python
if token.startswith('ey') and token.count('.') in (2, 4):
```

Reply:
@skion Done! Thanks.